### PR TITLE
Q-Learning algorithm implemented in Python

### PR DESCRIPTION
**Pull request by AryanRaj315**
### **Describe your change:**

* [x] Add an algorithm?
* [ ] Fix a bug or typo in an existing algorithm?
* [ ] Documentation change?

### **Checklist:**
* [x] I have read [CONTRIBUTING.md](https://github.com/TheAlgorithms/Python/blob/master/CONTRIBUTING.md).
* [x] This pull request is all my own work -- I have not plagiarized.
* [x] I know that pull requests will not be merged if they fail the automated tests.
* [x] This PR only changes one algorithm file.  To ease review, please open separate PRs for separate algorithms.
* [x] All new Python files are placed inside an existing directory.
* [x] All filenames are in all lowercase characters with no spaces or dashes.
* [ ] All functions and variable names follow Python naming conventions.
* [ ] All function parameters and return values are annotated with Python [type hints](https://docs.python.org/3/library/typing.html).
* [ ] All functions have [doctests](https://docs.python.org/3/library/doctest.html) that pass the automated testing.
* [x] All new algorithms have a URL in its comments that points to Wikipedia or other similar explanation.
* [ ] If this pull request resolves one or more open issues then the commit message contains `Fixes: #{$ISSUE_NO}`.

### CODE REVIEW

File: genetic_algorithm/qlearning.py [hunk 1]
```diff
@@ -0,0 +1,254 @@
...
+        action = randomAction()
+    return action
+
+
+def randomAction():
+    """
+    This function returns random action just in case the Qtable values return None.
+    """
+    action = np.random.choice([0, 1, 2, 3])
+    return action
```
Comment (cclauss) [inline genetic_algorithm/qlearning.py:32]:
```suggestion
def next_best_action(state: int, Q_table: np.ndarray) -> int:
    """
    This function returns the most suitable action value from the qtable.
    
    >>> next_best_action(1, []) in [0, 1, 2, 3]
    True
    """
    action = np.argmax(Q_table[state])
    return action if action is not None else np.random.choice([0, 1, 2, 3])
```
Adds a proper function name in snake_case with Python type hints and a doctest.  Please read CONTRIBUTING.md.

**Comment by cclauss** [comment]
[FileNotFoundError: [Errno 2] No such file or directory: 'Qtable.txt'](https://travis-ci.com/github/TheAlgorithms/Python/builds/179843622#L408-L410)

File: genetic_algorithm/qlearning.py [hunk 2]
```diff
@@ -7,40 +7,36 @@
...
-    """
-    This function returns random action just in case the Qtable values return None.
-    """
-    action = np.random.choice([0, 1, 2, 3])
-    return action
+    return action if action is not None else np.random.choice([0, 1, 2, 3])
 
 
-def StateActionReward(player, x_food, y_food):
+def state_action_reward(player, x_food, y_food):
```
Comment (cclauss) [inline genetic_algorithm/qlearning.py:31]: Type hints?

File: genetic_algorithm/snake.py [hunk 3]
```diff
@@ -0,0 +1,239 @@
+import random
```
Comment (cclauss) [inline genetic_algorithm/snake.py]: One Python file per PR please.

File: genetic_algorithm/qlearning.py [hunk 4]
```diff
@@ -0,0 +1,248 @@
...
+import pickle
+
+GAMMA = 0.9
+LEARNING_RATE = 0.7
+
+with open("Qtable.txt", "rb") as f:
+    """
+    This function is responsible for loading Qtable.txt if already present
+    """
+    Q_table = pickle.load(f)
```
Comment (cclauss) [inline genetic_algorithm/qlearning.py:17]:
```suggestion
if __name__ == "__main__":
    with open("Qtable.txt", "rb") as f:
        """
        This code is responsible for loading Qtable.txt if already present
        """
        Q_table = pickle.load(f)
```

File: genetic_algorithm/qlearning.py [hunk 5]
```diff
@@ -0,0 +1,248 @@
...
+        or (player.dirny == 1 and y_agent == y_food and x_food < x_agent)
+        or (player.dirny == -1 and y_agent == y_food and x_food > x_agent)
+    )
+    # Adding to states list while priortizing danger over eating food
+    states = [
+        1 if DangerAhead is True else 0,
+        1 if DangerLeft is True else 0,
+        1 if DangerRight is True else 0,
+    ]
+    if (DangerAhead == 0) and (DangerRight == 0) and (DangerLeft == 0):
```
Comment (cclauss) [inline genetic_algorithm/qlearning.py:153]:
```suggestion
    if sum(states) == 0:
```

File: genetic_algorithm/qlearning.py [hunk 6]
```diff
@@ -0,0 +1,248 @@
...
+        or (player.dirnx == -1 and y_agent > y_food and x_food == x_agent)
+        or (player.dirny == 1 and y_agent == y_food and x_food < x_agent)
+        or (player.dirny == -1 and y_agent == y_food and x_food > x_agent)
+    )
+    # Adding to states list while priortizing danger over eating food
+    states = [
+        1 if DangerAhead is True else 0,
+        1 if DangerLeft is True else 0,
+        1 if DangerRight is True else 0,
+    ]
```
Comment (cclauss) [inline genetic_algorithm/qlearning.py:152]:
```suggestion
    states = [int(x) for x in (DangerAhead, DangerLeft, DangerRight)]
```

File: genetic_algorithm/qlearning.py [hunk 7]
```diff
@@ -0,0 +1,248 @@
...
+    if (DangerAhead == 0) and (DangerRight == 0) and (DangerLeft == 0):
+        states.append(1 if FoodStraightAhead is True else 0)
+        states.append(1 if FoodAheadRight is True else 0)
+        states.append(1 if FoodAheadLeft is True else 0)
+        states.append(1 if FoodBehindRight is True else 0)
+        states.append(1 if FoodBehindLeft is True else 0)
+        states.append(1 if FoodBehind is True else 0)
+        states.append(1 if FoodLeft is True else 0)
+        states.append(1 if FoodRight is True else 0)
```
Comment (cclauss) [inline genetic_algorithm/qlearning.py:161]:
```suggestion
        states += [int(x) for x in (
            FoodStraightAhead, FoodAheadRight, FoodAheadLeft, FoodBehindRight,
            FoodBehindLeft, FoodBehind, FoodLeft, FoodRight
        )]
```

File: genetic_algorithm/qlearning.py [hunk 8]
```diff
@@ -0,0 +1,248 @@
...
+    else:
+        for i in range(8):
+            states.append(0)
```
Comment (cclauss) [inline genetic_algorithm/qlearning.py:164]:
```suggestion
        states += [0] * 8
```

File: genetic_algorithm/qlearning.py [hunk 9]
```diff
@@ -0,0 +1,248 @@
...
+    state = 0
+    for i in range(11):
+        if states[i] == 1:
+            state = i
+    return state
+
+
+def dist(x1, y1, x2, y2):
```
Comment (cclauss) [inline genetic_algorithm/qlearning.py:172]:
```suggestion
def euler_distance(x1, y1, x2, y2):
```

File: genetic_algorithm/qlearning.py [hunk 10]
```diff
@@ -0,0 +1,248 @@
...
+
+def learn(state, action, reward, next_state, next_action, i, trialNumber, epsilon):
+    """
+    This function is for iteratively updating the Qtable.
+    """
+    currentQ = Q_table[state][action]
+    nextQ = Q_table[next_state][next_action]
+    newQ = (1 - LEARNING_RATE) * currentQ + LEARNING_RATE * (
+        reward + GAMMA * nextQ
+    )  # Qlearning Algorithm to get new q value for the q table.
```
Comment (cclauss) [inline genetic_algorithm/qlearning.py:240]:
```suggestion
    # Qlearning Algorithm to get new q value for the q table.
    newQ = (1 - LEARNING_RATE) * currentQ + LEARNING_RATE * (reward + GAMMA * nextQ)  
```

File: genetic_algorithm/qlearning.py [hunk 11]
```diff
@@ -0,0 +1,248 @@
...
+    Q_table[state][action] = newQ
+    state = next_state
+    currentQ = nextQ
+    if trialNumber % 100 == 0:
+        print("Printing Q_table: ")
+        print(Q_table)
```
Comment (cclauss) [inline genetic_algorithm/qlearning.py:246]:
```suggestion
        print(f"{Q_table = }")
```

Comment (cclauss) [reply genetic_algorithm/qlearning.py:246]: Requires Python >= 3.8

File: genetic_algorithm/qlearning.py [hunk 12]
```diff
@@ -0,0 +1,248 @@
...
+# States to consider:
+#   * Food's relative positioning wrt the head
+# Checking for food in nine directions
+###
+# #
+###
+#   * obstruction Ahead, Right, Left
+
+
+def state(player, x_food, y_food):
```
Comment (cclauss) [inline genetic_algorithm/qlearning.py:51]: Type hints?

File: genetic_algorithm/qlearning.py (hunk 2 above)
Comment (AryanRaj315) [reply genetic_algorithm/qlearning.py:26]: Making changes. Thank you for your help!

**Comment by cclauss** [comment]
This repo does not accept Jupyter notebooks but https://github.com/TheAlgorithms/Jupyter does.

**Pull request closed without merging**